### PR TITLE
Feature/transport configuration rework

### PR DESCRIPTION
**Pull request by chrisdutz**
Reworked the configuration of transports so a transports configuration options are referenced by a prefix, that matches the transport code.

So for example in S7 I can customize the timeout like this:

    s7://192.168.23.30?tcp.default-timeout=3000

### CODE REVIEW

File: plc4j/drivers/ads/pom.xml [hunk 1]
```diff
@@ -140,11 +140,6 @@
       <artifactId>plc4j-transport-tcp</artifactId>
       <version>0.12.0-SNAPSHOT</version>
     </dependency>
-    <dependency>
-      <groupId>org.apache.plc4x</groupId>
-      <artifactId>plc4j-transport-serial</artifactId>
```
Comment (sruehl) [inline plc4j/drivers/ads/pom.xml:103]: how is this related to this PR? or is that just a simple cleanup?

File: plc4j/drivers/ab-eth/src/main/java/org/apache/plc4x/java/abeth/configuration/AbEthConfiguration.java [hunk 2]
```diff
@@ -18,12 +18,10 @@
  */
 package org.apache.plc4x.java.abeth.configuration;
 
-import org.apache.plc4x.java.abeth.AbEthDriver;
 import org.apache.plc4x.java.spi.configuration.Configuration;
 import org.apache.plc4x.java.spi.configuration.annotations.ConfigurationParameter;
-import org.apache.plc4x.java.transport.tcp.TcpTransportConfiguration;
 
-public class AbEthConfiguration implements Configuration, TcpTransportConfiguration {
```
Comment (sruehl) [inline plc4j/drivers/ab-eth/src/main/java/org/apache/plc4x/java/abeth/configuration/AbEthConfiguration.java:26]: instead of separating it you could have just added extends `DefaultTcpTransportConfiguration` but I don't see a issue in having a separate class `AbEthTcpTransportConfiguration`

File: plc4j/drivers/ads/src/main/java/org/apache/plc4x/java/ads/configuration/AdsConfiguration.java [hunk 3]
```diff
@@ -19,22 +19,17 @@
...
 import org.apache.plc4x.java.spi.configuration.annotations.defaults.BooleanDefaultValue;
 import org.apache.plc4x.java.spi.configuration.annotations.defaults.IntDefaultValue;
-import org.apache.plc4x.java.transport.serial.SerialTransportConfiguration;
-import org.apache.plc4x.java.transport.tcp.TcpTransportConfiguration;
 
 import java.util.regex.Pattern;
 import java.util.stream.Stream;
 
-public class AdsConfiguration implements Configuration, TcpTransportConfiguration, SerialTransportConfiguration {
+public class AdsConfiguration implements Configuration {
```
Comment (sruehl) [inline plc4j/drivers/ads/src/main/java/org/apache/plc4x/java/ads/configuration/AdsConfiguration.java:32]: ok guess it is because here you need to extend 2 types... In Golang this could be also done with delegates but in Java world it seem easier to just throw it in 2 classes 👍🏼

File: plc4j/drivers/ab-eth/src/main/java/org/apache/plc4x/java/abeth/configuration/AbEthConfiguration.java (hunk 2 above)
Comment (sruehl) [reply plc4j/drivers/ab-eth/src/main/java/org/apache/plc4x/java/abeth/configuration/AbEthConfiguration.java:26]: Guess it is for consistency... see comment below on the other class

File: plc4j/drivers/c-bus/src/main/java/org/apache/plc4x/java/cbus/configuration/CBusTcpTransportConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+package org.apache.plc4x.java.cbus.configuration;
+
+import org.apache.plc4x.java.transport.tcp.DefaultTcpTransportConfiguration;
+
+public class CBusTcpTransportConfiguration extends DefaultTcpTransportConfiguration {
+
+    @Override
+    public int getDefaultPort() {
+        return 123;
```
Comment (sruehl) [inline plc4j/drivers/c-bus/src/main/java/org/apache/plc4x/java/cbus/configuration/CBusTcpTransportConfiguration.java]: One thing noticing here. Guess this is a placeholder but there is a constant in the mspec for that

Comment (sruehl) [reply plc4j/drivers/c-bus/src/main/java/org/apache/plc4x/java/cbus/configuration/CBusTcpTransportConfiguration.java]: or should be....

File: plc4j/drivers/firmata/pom.xml [hunk 5]
```diff
@@ -122,12 +122,6 @@
       <version>0.12.0-SNAPSHOT</version>
     </dependency>
 
-    <dependency>
-      <groupId>org.apache.plc4x</groupId>
-      <artifactId>plc4j-transport-serial</artifactId>
```
Comment (sruehl) [inline plc4j/drivers/firmata/pom.xml:106]: same here, why is the dependency disappearing

File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/CustomProtocolStackConfigurer.java [hunk 6]
```diff
@@ -101,7 +102,7 @@ public Plc4xProtocolBase<BASE_PACKET_CLASS> configurePipeline(Configuration conf
         if (driverContext != null) {
             protocol.setDriverContext(driverContext);
         }
-        Plc4xNettyWrapper<BASE_PACKET_CLASS> context = new Plc4xNettyWrapper<>(configuration.getTimeoutManager(), pipeline, passive, protocol, authentication, basePacketClass);
+        Plc4xNettyWrapper<BASE_PACKET_CLASS> context = new Plc4xNettyWrapper<>(new NettyHashTimerTimeoutManager(), pipeline, passive, protocol, authentication, basePacketClass);
```
Comment (sruehl) [inline plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/CustomProtocolStackConfigurer.java:105]: what is this change about?

File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/GeneratedDriverBase.java [hunk 7]
```diff
@@ -150,8 +153,24 @@ public PlcConnection getConnection(String connectionString, PlcAuthentication au
...
-        // Inject the configuration into the transport.
-        configure(configuration, transport);
+        // Find out the type of the transport configuration.
+        Class<? extends TransportConfiguration> transportConfigurationType = transport.getTransportConfigType();
+        if(this instanceof TransportConfigurationTypeProvider) {
+            TransportConfigurationTypeProvider transportConfigurationTypeProvider =
+                (TransportConfigurationTypeProvider) this;
+            Class<? extends TransportConfiguration> driverTransportConfigurationType =
+                transportConfigurationTypeProvider.getTransportConfigurationType(transportCode);
+            if(driverTransportConfigurationType != null) {
```
Comment (sruehl) [inline plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/GeneratedDriverBase.java:163]: maybe some trace/debug level logging might be helpful here at some point

File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/SingleProtocolStackConfigurer.java [hunk 8]
```diff
@@ -111,7 +112,7 @@ public Plc4xProtocolBase<BASE_PACKET_CLASS> configurePipeline(Configuration conf
         if (driverContextClass != null) {
             protocol.setDriverContext(configure(configuration, createInstance(driverContextClass)));
         }
-        Plc4xNettyWrapper<BASE_PACKET_CLASS> context = new Plc4xNettyWrapper<>(configuration.getTimeoutManager(), pipeline, passive, protocol,
+        Plc4xNettyWrapper<BASE_PACKET_CLASS> context = new Plc4xNettyWrapper<>(new NettyHashTimerTimeoutManager(), pipeline, passive, protocol,
```
Comment (sruehl) [inline plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/SingleProtocolStackConfigurer.java:115]: same as above: why is this change?

File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/transport/Transport.java [hunk 9]
```diff
@@ -28,4 +28,8 @@ public interface Transport {
 
     ChannelFactory createChannelFactory(String transportConfig);
 
+    default Class<? extends TransportConfiguration> getTransportConfigType() {
+        return null;
```
Comment (sruehl) [inline plc4j/spi/src/main/java/org/apache/plc4x/java/spi/transport/Transport.java:32]: we could also consider the use of optional, depending on how it is use (not a issue with null)

File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/transport/TransportConfigurationTypeProvider.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.plc4x.java.spi.transport;
+
+public interface TransportConfigurationTypeProvider {
+
+    Class<? extends TransportConfiguration> getTransportConfigurationType(String transportCode);
```
Comment (sruehl) [inline plc4j/spi/src/main/java/org/apache/plc4x/java/spi/transport/TransportConfigurationTypeProvider.java:24]: see optional comment above

File: plc4j/spi/src/test/java/org/apache/plc4x/java/spi/configuration/ConfigurationFactoryTest.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+ */
+
+package org.apache.plc4x.java.spi.configuration;
+
+import org.apache.plc4x.java.spi.configuration.config.*;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ConfigurationFactoryTest {
```
Comment (sruehl) [inline plc4j/spi/src/test/java/org/apache/plc4x/java/spi/configuration/ConfigurationFactoryTest.java:27]: nice, tests... :)

File: plc4j/transports/pcap-shared/src/main/java/org/apache/plc4x/java/transport/pcap/PcapTransportConfiguration.java [hunk 12]
```diff
@@ -19,22 +19,18 @@
...
+    int NO_DEFAULT_PORT = -1;
 
-    default boolean getSupportVlans() {
-        return false;
-    }
+    boolean getSupportVlans();
 
     default int getDefaultPort() {
-        return PcapChannelConfig.ALL_PORTS;
+        return NO_DEFAULT_PORT;
```
Comment (sruehl) [inline plc4j/transports/pcap-shared/src/main/java/org/apache/plc4x/java/transport/pcap/PcapTransportConfiguration.java:30]: why is this logic change here?

**Review by sruehl — APPROVED** [review]

**Comment by sruehl** [comment]
LGTM

File: plc4j/drivers/firmata/pom.xml (hunk 5 above)
Comment (chrisdutz) [reply plc4j/drivers/firmata/pom.xml:106]: Because we no longer directly reference serial as we rely on the defaults ...

File: plc4j/drivers/c-bus/src/main/java/org/apache/plc4x/java/cbus/configuration/CBusTcpTransportConfiguration.java (hunk 4 above)
Comment (chrisdutz) [reply plc4j/drivers/c-bus/src/main/java/org/apache/plc4x/java/cbus/configuration/CBusTcpTransportConfiguration.java]: Well ... before refactoring it seems someone had "123" before ... I just took that over.

File: plc4j/drivers/ads/pom.xml (hunk 1 above)
Comment (chrisdutz) [reply plc4j/drivers/ads/pom.xml:103]: Well after the refactoring, everything using the serial no longer explicitly imported any of the types of the serial module. We're now using the normal defaults from the serial transport. So if I hadn't removed it, the dependncy plugin would have failed the build.

File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/SingleProtocolStackConfigurer.java (hunk 8 above)
Comment (chrisdutz) [reply plc4j/spi/src/main/java/org/apache/plc4x/java/spi/connection/SingleProtocolStackConfigurer.java:115]: This was simply me cleaning up something that I thought didn't belong into the configuration.

File: plc4j/transports/pcap-shared/src/main/java/org/apache/plc4x/java/transport/pcap/PcapTransportConfiguration.java (hunk 12 above)
Comment (chrisdutz) [reply plc4j/transports/pcap-shared/src/main/java/org/apache/plc4x/java/transport/pcap/PcapTransportConfiguration.java:30]: Well ... this way it matches the constant where it's used in RawSocketTransport.java

File: plc4j/drivers/c-bus/src/main/java/org/apache/plc4x/java/cbus/configuration/CBusTcpTransportConfiguration.java (hunk 4 above)
Comment (sruehl) [reply plc4j/drivers/c-bus/src/main/java/org/apache/plc4x/java/cbus/configuration/CBusTcpTransportConfiguration.java]: fixed it
